Removed temporary memory from the API

slouken opened the pull request:
It was intended to make the API easier to use, but various automatic garbage collection all had flaws, and making the application periodically clean up temporary memory added cognitive load to using the API, and in many cases was it was difficult to restructure threaded code to handle this.

So, we're largely going back to the original system, where the API returns allocated results and you free them.

In addition, to solve the problems we originally wanted temporary memory for:
* Short strings with a finite count, like device names, get stored in a per-thread string pool.
* Events continue to use temporary memory internally, which is cleaned up on the next event processing cycle.

slouken commented:
@icculus, @sezero, this was a bunch of code changes, and I guarantee there are bugs here. See if you can spot them? :D

slouken commented:
Also, I have an sdl2-compat change coming...

icculus commented on include/SDL3/SDL_begin_code.h:
```diff
@@ -227,3 +225,8 @@
+
+/* This is used to mark functions that return memory that need to be freed with SDL_free() */
+#ifndef SDL_DECLSPEC_FREE
+#define SDL_DECLSPEC_FREE   SDL_DECLSPEC SDL_MALLOC
```
This isn't what SDL_MALLOC does. It doesn't mean "this allocates some memory that needs to be freed," but rather "this is an _allocator_ function like malloc with certain rules." Among them, that the allocated memory is _uninitialized_ and the compiler can optimize for that truth.

At least according to https://stackoverflow.com/questions/18485447/gcc-attribute-malloc

slouken replied:
Oh, okay, fixing shortly.

icculus commented:
Yeah, let's merge this.

What a long, strange trip this has been.

slouken commented:
> Yeah, let's merge this.
> 
> What a long, strange trip this has been.

Oy...

sezero commented:
MinGW x86 tests in CI have been failing after this: https://github.com/libsdl-org/SDL/actions/runs/10120759219/job/27990895461

madebr commented:
- I think there's an issue with windows (clang-cl), but it does not always reproduce: https://github.com/libsdl-org/SDL/actions/runs/10122580945/job/27994890685#step:9:19888 (stdlib_getsetenv)
- `testsprite` example does not show any sprite (Linux), testspriteminimal works
- `testwm` does not show any text
- `testaudio --trackmem`  shows a leak (Linux) (and `--trackmem` ordens the icons differently but that's not a real issue)
- `testyuv --trackmem` shows a leak through `SDL_GetBasePath` (`SDL_GetBasePath` currently returns `const char *`)
- I know SDL_test's memory tracker is extremely inefficient, but `testgles2 --trackmem` and `testrendertarget --trackmem` are very slow at startup and shows intermediate slowdowns

slouken commented:
> MinGW x86 tests in CI have been failing after this: https://github.com/libsdl-org/SDL/actions/runs/10120759219/job/27990895461

Yep, I'm working on this now.

slouken commented:
> * I think there's an issue with windows (clang-cl), but it does not always reproduce: https://github.com/libsdl-org/SDL/actions/runs/10122580945/job/27994890685#step:9:19888 (stdlib_getsetenv)

I think what's happening here is that somewhere we're freeing a const string and returning undefined memory. I'm going to see if I can track this down.

> * `testsprite` example does not show any sprite (Linux), testspriteminimal works
> * `testwm` does not show any text

Fixed

> * `testaudio --trackmem`  shows a leak (Linux) (and `--trackmem` ordens the icons differently but that's not a real issue)

Fixed

> * `testyuv --trackmem` shows a leak through `SDL_GetBasePath` (`SDL_GetBasePath` currently returns `const char *`)

Fixed

> * I know SDL_test's memory tracker is extremely inefficient, but `testgles2 --trackmem` and `testrendertarget --trackmem` are very slow at startup and shows intermediate slowdowns

I'm not seeing the slowdown here, are you able to break at startup and see where it's spending time?

slouken commented:
> > * I think there's an issue with windows (clang-cl), but it does not always reproduce: https://github.com/libsdl-org/SDL/actions/runs/10122580945/job/27994890685#step:9:19888 (stdlib_getsetenv)
> 
> I think what's happening here is that somewhere we're freeing a const string and returning undefined memory. I'm going to see if I can track this down.

This should be fixed in https://github.com/libsdl-org/SDL/commit/9da58e9fb7014a6b069433a1484198032602fda4

madebr commented:
> I'm not seeing the slowdown here, are you able to break at startup and see where it's spending time?

I ran gprof and ~99% of the time is in SDLTest's `SDL_TrackAllocation`.

Moving the symbol name formatting to `SDLTest_LogAllocations` speeds things up considerably.
The issue with this is that leaks in unloaded modules won't get resolved.

madebr commented:
<details>

<summary>Leaks of testautomaton on Linux</summary>

```
Allocation 0: 72 bytes
	0x7fab29518055: SDL_realloc_REAL+0x43
	0x7fab2951daa0: SDL_SetTLS_REAL+0xff
	0x7fab2951e071: SDL_GetErrBuf+0xc9
	0x7fab294325e3: SDL_SetError_REAL+0x77
	0x7fab2951f9fe: SDL_RemoveTimer_REAL+0x13f
	0x7fab29469e3a: SDL_RemoveTimer+0x1f
	0x458716: SDLTest_RunTest+0x19b
	0x45908b: SDLTest_RunSuites+0x83d
	0x40a254: main+0x4ec
	0x7fab2922c14a: __libc_start_call_main+0x7a
	0x7fab2922c20b: __libc_start_main+0x8b
	0x409c35: _start+0x25
Allocation 1: 40 bytes
	0x7fab29517fe6: SDL_calloc_REAL+0x4b
	0x7fab296de775: SDL_CreateMutex_REAL+0x1d
	0x7fab29434af6: SDL_LogMessageV_REAL+0x88
	0x7fab2946f026: SDL_LogMessage+0xb0
	0x45969b: SDLTest_Log+0xfe
	0x44d719: SDLTest_AssertPass+0xf3
	0x43eea2: subsystemsTearDown+0x25
	0x458706: SDLTest_RunTest+0x18b
	0x45908b: SDLTest_RunSuites+0x83d
	0x40a254: main+0x4ec
	0x7fab2922c14a: __libc_start_call_main+0x7a
	0x7fab2922c20b: __libc_start_main+0x8b
	0x409c35: _start+0x25
```

</details>

<details>

<summary>Leaks of testsprite on wine (Windows)</summary>

```
Allocation 0: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa3012a0: D3D11_CreateRenderer+0x357 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x1400083d7: SDLTest_CommonInit+0x1263 :0
        0x14000371f: SDL_AppInit+0x639 :0
        0x6ffffa2e3800: SDL_InitMainCallbacks+0x69 :0
        0x6ffffa505366: SDL_EnterAppMainCallbacks_REAL+0x3f :0
        0x6ffffa2a3732: SDL_EnterAppMainCallbacks+0x44 :0
        0x140001843: SDL_main+0x43 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x14000187d: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 1: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa30a0c2: D3D12_CreateRenderer+0x381 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x1400083d7: SDLTest_CommonInit+0x1263 :0
        0x14000371f: SDL_AppInit+0x639 :0
        0x6ffffa2e3800: SDL_InitMainCallbacks+0x69 :0
        0x6ffffa505366: SDL_EnterAppMainCallbacks_REAL+0x3f :0
        0x6ffffa2a3732: SDL_EnterAppMainCallbacks+0x44 :0
        0x140001843: SDL_main+0x43 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x14000187d: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Total: 0.08 Kb in 2 allocations
```

</details>


Also, the `render_testBlendModes` fails on wine:
```
```
<details>

<summary>Leaks of testautomation on wine (Windows)</summary>

```
INFO: Memory allocations:
Allocation 0: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa3012a0: D3D11_CreateRenderer+0x357 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 1: 72 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa376923: SDL_SetTLS_REAL+0xf4 :0
        0x6ffffa376ee5: SDL_GetErrBuf+0xc4 :0
        0x6ffffa2729dd: SDL_SetError_REAL+0x2d :0
        0x6ffffa3787cc: SDL_RemoveTimer_REAL+0x12f :0
        0x6ffffa2a8479: SDL_RemoveTimer+0x19 :0
        0x140053836: SDLTest_RunTest+0x1aa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 2: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa3012a0: D3D11_CreateRenderer+0x357 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 3: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa3012a0: D3D11_CreateRenderer+0x357 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x140048276: video_getWindowSurface+0x1e9 :0
        0x1400537d2: SDLTest_RunTest+0x146 :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 4: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa30a0c2: D3D12_CreateRenderer+0x381 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 5: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa30a0c2: D3D12_CreateRenderer+0x381 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 6: 16 bytes
        0x6ffffa370cf4: SDL_calloc_REAL+0x43 :0
        0x6ffffa4c9507: SDL_CreateMutex_srw+0x17 :0
        0x6ffffa4c99b8: SDL_CreateMutex_REAL+0x139 :0
        0x6ffffa274ad1: SDL_LogMessageV_REAL+0x8a :0
        0x6ffffa2ad141: SDL_LogMessage+0x3e :0
        0x14005484a: SDLTest_Log+0xad :0
        0x14004860d: SDLTest_AssertPass+0xa2 :0
        0x140039f07: subsystemsTearDown+0x20 :0
        0x140053826: SDLTest_RunTest+0x19a :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 7: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa30a0c2: D3D12_CreateRenderer+0x381 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 8: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa30a0c2: D3D12_CreateRenderer+0x381 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 9: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa3012a0: D3D11_CreateRenderer+0x357 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 10: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa30a0c2: D3D12_CreateRenderer+0x381 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 11: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa3012a0: D3D11_CreateRenderer+0x357 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 12: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa3012a0: D3D11_CreateRenderer+0x357 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 13: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa3012a0: D3D11_CreateRenderer+0x357 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 14: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa30a0c2: D3D12_CreateRenderer+0x381 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 15: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa3012a0: D3D11_CreateRenderer+0x357 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 16: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa3012a0: D3D11_CreateRenderer+0x357 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 17: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa3012a0: D3D11_CreateRenderer+0x357 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 18: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa30a0c2: D3D12_CreateRenderer+0x381 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x140048276: video_getWindowSurface+0x1e9 :0
        0x1400537d2: SDLTest_RunTest+0x146 :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 19: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa30a0c2: D3D12_CreateRenderer+0x381 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 20: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa30a0c2: D3D12_CreateRenderer+0x381 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 21: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa30a0c2: D3D12_CreateRenderer+0x381 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 22: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa30a0c2: D3D12_CreateRenderer+0x381 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 23: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa3012a0: D3D11_CreateRenderer+0x357 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 24: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa3012a0: D3D11_CreateRenderer+0x357 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 25: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa30a0c2: D3D12_CreateRenderer+0x381 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 26: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa3012a0: D3D11_CreateRenderer+0x357 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Allocation 27: 40 bytes
        0x6ffffa370d5f: SDL_realloc_REAL+0x3b :0
        0x6ffffa2e44a6: SDL_AddSupportedTextureFormat+0x39 :0
        0x6ffffa30a0c2: D3D12_CreateRenderer+0x381 :0
        0x6ffffa2e67d2: SDL_CreateRendererWithProperties_REAL+0x28e :0
        0x6ffffa2e6c9d: SDL_CreateRenderer_REAL+0x52 :0
        0x6ffffa2a2fec: SDL_CreateRenderer+0x25 :0
        0x14002a1e4: InitCreateRenderer+0xa4 :0
        0x140053786: SDLTest_RunTest+0xfa :0
        0x1400541b5: SDLTest_RunSuites+0x82e :0
        0x1400019f5: SDL_main+0x535 :0
        0x6ffffa4ad112: SDL_RunApp_REAL+0x26e :0
        0x6ffffa29d62e: SDL_RunApp_DEFAULT+0x3d :0
        0x6ffffa2a8cb2: SDL_RunApp+0x38 :0
        0x140001484: WinMain+0x34 :0
        0x1400012ee: ???+0x0 :0
        0x140001406: ???+0x0 :0
        0x6fffffa88c79: BaseThreadInitThunk+0x9 :0
        0x6fffffc86e8f: ???+0x0 :0
Total: 1.10 Kb in 28 allocations
```

</details>
